### PR TITLE
Set backupskip default for macOS

### PR DESCRIPTION
**Pull request by raimue**
On macOS, /tmp is actually a symlink to /private/tmp. This change is necessary to ensure that 'crontab -e' works as expected.

Note that this is only a problem when building your own vim. For /usr/bin/vim, an autocommand in /usr/share/vim/vimrc sets 'nobackup' for `/private/tmp/crontab.*`, but of course this will not be loaded when installing to a different prefix.

### CODE REVIEW

**Comment by nuko8** [comment]
> This change is necessary to ensure that 'crontab -e' works as expected.

> Note that this is only a problem when building your own vim.

As far as `crontab -e` is concerned, those two statements sound too strong.  Note that `/usr/bin/vi` is hard-coded in `/usr/bin/crontab`, although that fact is not documented in `crontab(1)`.  Also, `/usr/bin/vi` is a symbolic link to `/usr/bin/vim`.  Accordingly, if both of the environment variables `VISUAL` and `EDITOR` are undefined or `unset` when `crontab -e` is invoked (cf. `crontab(1)`), `/usr/bin/vim` tailored by `/usr/share/vim/vimrc` for `crontab -e` will be called exclusively even if there's another vim installed somewhere else, and there should be no trouble with that.  In other words, the default works fine enough, and having  a user's own vim won't prevent `crontab -e` from working unless `crontab` is directed to use a vim that doesn't take care of the `crontab -e` usage on macOS.

**Comment by raimue** [comment]
Well, you described the problem exactly. `crontab -e` will launch `$EDITOR` from the environment. This is a very common configuration and causes problems that can be avoided by initializing `backupskip` with a better default.

```
export EDITOR=~/bin/vim
crontab -e
```

`crontab -e` is just one use case, having `/tmp` in that list is not useful on macOS at all. As it is a symlink to `/private/tmp`, no file will ever match against `/tmp/*`.

**Comment by nuko8** [comment]
> having /tmp in that list is not useful on macOS at all. As it is a symlink to /private/tmp, no file will ever match against /tmp/*.

So the proposed change of this PR is worthy to be included for its own sake, isn't it?

If that happens, the description of `'backupskip'` (`runtime/doc/options.txt`) should be updated accordingly.  Giving the patch a brief comment on why macOS needs `/private/tmp` instead of `/tmp`, and adding a test on the proposed `'bsk'` (See `src/testdir/{README.txt,opt_test.vim}`), would be welcomed.

**Comment by nuko8** [comment]
In addition, please consider using `MACOS_X` instead of `__APPLE__` for the sake of consistency: https://github.com/vim/vim/blob/9ce4213b0322477ace1df5e0aa356b1a5551b445/src/vim.h#L93

**Comment by raimue** [comment]
I do not understand what you mean with "for its own sake"... Yes, this needs to be fixed otherwise I would not open a pull request?

Thanks for pointing me to `MACOS_X`. Why this change is necessary was already in the commit message, but I expanded on it.

Not sure what kind of test you would expect. A check that the default value was set differently would just be redundant.

**Comment by chrisbra** [comment]
Does that apply to all Mac OS X versions? Or are there versions where /tmp is no link?

> Not sure what kind of test you would expect. A check that the default value was set differently would just be redundant.

Perhaps something like this?
```patch
diff --git a/src/testdir/test_options.vim b/src/testdir/test_options.vim
index 5d2033a5d..d46a3439d 100644
--- a/src/testdir/test_options.vim
+++ b/src/testdir/test_options.vim
@@ -332,3 +332,16 @@ func Test_set_indentexpr()
   call assert_equal('', &indentexpr)
   bwipe!
 endfunc
+
+func Test_backupskip_mac()
+  if has("mac")
+    call assert_equal('/private/tmp/*', &bsk)
+  elseif has("win32")
+    let list=['$TEMPDIR','$TMP', '$TEMP']
+               let expected=map(copy(list), {c,v -> exists(v) ? expand(v).'\*':''})
+               call filter(expected, {c,v -> !empty(v)})
+               call assert_equal(join(expected, ','), &bsk)
+  elseif has("unix")
+    call assert_equal('/tmp/*', &bsk)
+  endif
+endfunc
```

**Comment by chrisbra** [comment]
Note this is untested.

File: runtime/doc/options.txt [hunk 1]
```diff
@@ -1087,7 +1087,10 @@ A jump table for the options with a short description can be found at |Q_op|.
 <	Use 'backupdir' to put the backup in a different directory.
 
 						*'backupskip'* *'bsk'*
-'backupskip' 'bsk'	string	(default: "/tmp/*,$TMPDIR/*,$TMP/*,$TEMP/*")
+'backupskip' 'bsk'	string	(default: "$TMPDIR/*,$TMP/*,$TEMP/*",
+				 Unix: "/tmp/*,$TMPDIR/*,$TMP/*,$TEMP/*",
+				 Mac OS X:
+				 "/private/tmp/*,$TMPDIR/*,$TMP/*,$TEMP/*")
```
Comment (nuko8) [inline runtime/doc/options.txt:1093]: Please explain why you made more changes than adding the macOS entry.  Even if the changes appear to be more consistent with the description of `:h bsk`, people probably want to know why the value of `default` was changed and the new entry `Unix` was introduced.

Or,

1. Remove the whole line of `default`,
2. Make `Unix` revert to `default`, and
3. Change `Mac OS X` to `macOS`.

Comment (nuko8) [reply runtime/doc/options.txt:1093]: Ah, okay.  Bram already approved the changes.  Since I've been away from the keyboard since yesterday evening until now, I didn't notice the patch has already been included.  Sorry about the noise.

**Comment by nuko8** [comment]
> Does that apply to all Mac OS X versions? Or are there versions where /tmp is no link?

Unfortunately, I don't have a clear answer to those questions and couldn't find a nice one on the Internet.

What I can tell at this moment is:

1.  As a user since 10.6 (Snow Leopard, released on August 28, 2009), I know `/tmp/` has always been a symbolic link to `/private/tmp/` since 10.6.

2. Having googled, I found out an excerpt of a published book which says that 10.3 (Panther, released on October 24, 2003) has the symbolic link in question.


3. ~~~According to `hier(7)` (dated July 23, 2007),~~~
```man
NAME
     hier -- layout of filesystems

DESCRIPTION
...
/tmp/         temporary files
...
/var/         multi-purpose log, temporary, transient, and spool files
...
      tmp/       temporary files that are kept between system
                 reboots
```
~~~This is part of "BSD Miscellaneous Information Manual,"  and the descriptions of `/tmp/` and `/private/tmp/` indicate that each of them once played a role different from another, which doesn't apply to recent versions since such casting is impossible if they are linked.  Accordingly, there was seemingly a period of time when `/tmp/` wasn't a symbolic link of `/private/tmp/`; at worst, that could be the case for 10.2 (Jaguar, released on August 23, 2002).~~~

In conclusion, I think we are on a fairly safer side concerning the version difference issue.

**Comment by nuko8** [comment]
My bad.  Please ignore 3; somehow I got confused about the names between `/var/` and `/private/`.

As `hier(7)` doesn't mention `/private/` at all, I visited developer.apple.com and skimmed [File System Programming Guide](https://developer.apple.com/library/content/documentation/FileManagement/Conceptual/FileSystemProgrammingGuide/Introduction/Introduction.html#//apple_ref/doc/uid/TP40010672-CH1-SW1), and...to my  surprise, I couldn't find out `/private/` or any clue to it there.  Seems like that's a place where even devs are kept out...literally private EOL

**Comment by chrisbra** [comment]
Seems we are safe here. Thanks for digging through the Internet.
